action/recipes/linux_test.go

```diff
@@ -0,0 +1,169 @@
+package recipes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"dagger.io/dagger"
+	"github.com/9elements/firmware-action/action/container"
+	"github.com/9elements/firmware-action/action/filesystem"
+	"github.com/Masterminds/semver"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLinux(t *testing.T) {
+	// This test is really slow (like 100 seconds)
+	/*
+		if testing.Short() {
+			t.Skip("skipping test in short mode")
+		}
+	*/
+	pwd, err := os.Getwd()
+	assert.NoError(t, err)
+	defer os.Chdir(pwd) // nolint:errcheck
+
+	// Use "" if you want to test containers from github package registry
+	// Use "../../docker/linux" if you want to test containers built fresh from Dockerfile
+	dockerfilePath := ""
+	if true {
+		dockerfilePath, err = filepath.Abs("../../docker/linux")
+		assert.NoError(t, err)
+	}
+
+	testCases := []struct {
+		name         string
+		linuxVersion string
+		arch         string
+		wantErr      error
+	}{
+		{
+			name:         "normal build for x86 64bit",
+			linuxVersion: "6.1.45",
+			arch:         "x86_64",
+			wantErr:      nil,
+		},
+		{
+			name:         "normal build for x86 32bit",
+			linuxVersion: "6.1.45",
+			arch:         "x86",
+			wantErr:      nil,
+		},
+		{
+			name:         "normal build for arm64",
+			linuxVersion: "6.1.45",
+			arch:         "arm64",
+			wantErr:      nil,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.NoError(t, os.Chdir(pwd)) // just to make sure
+
+			linuxVersion, err := semver.NewVersion(tc.linuxVersion)
+			assert.NoError(t, err)
+			ctx := context.Background()
+			client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
+			assert.NoError(t, err)
+			defer client.Close()
+
+			// Prepare options
+			tmpDir := t.TempDir()
+			opts := map[string]string{
+				"target":           "linux",
+				"sdk_version":      fmt.Sprintf("linux_%s:main", linuxVersion.String()),
+				"architecture":     tc.arch,
+				"repo_path":        filepath.Join(tmpDir, "linux"),
+				"defconfig_path":   "custom_defconfig",
+				"containerWorkDir": "/linux",
+				"GITHUB_WORKSPACE": "/linux",
+				"output":           "output",
+			}
+			getFunc := func(key string) string {
+				return opts[key]
+			}
+			common, err := commonGetOpts(getFunc)
+			assert.NoError(t, err)
+			linuxOpts := linuxOpts{}
+
+			// Change current working directory
+			//   create __tmp_files__ directory to store source-code of Linux Kernel
+			//   mostly useful for repeated local-run tests to save bandwidth and time
+			tmpFiles := filepath.Join(pwd, "__tmp_files__")
+			err = exec.Command("mkdir", "-p", tmpFiles).Run()
+			assert.NoError(t, err)
+			err = os.Chdir(tmpFiles)
+			assert.NoError(t, err)
+
+			// Download linux source code to __tmp_files__
+			var commands [][]string
+			if errors.Is(filesystem.CheckFileExists(fmt.Sprintf("linux-%s", linuxVersion.String())), os.ErrNotExist) {
+				commands = [][]string{
```

Comment:
I think for now we can leave these tests to just run a bunch of external commands, ideally the tests should not be dependent on the outside OS as we shouldn't assume the users environment if they want to reproduce the test results. Both `wget` and `unxz` are easy to do in Go natively, for the `gpg2` stuff we'd have to check if a library already exists (probably it does). But let's keep that in mind for a later release.